src/doc.rs

```diff
@@ -568,35 +563,40 @@ impl EpubDoc {
     }
 }
 
-fn get_root_file(container: Vec<u8>) -> Result<String, Error> {
+fn get_root_file(container: Vec<u8>) -> Result<PathBuf, Error> {
     let xml = xmlutils::XMLReader::new(container.as_slice());
     let root = xml.parse_xml()?;
     let el = root.borrow();
     let element = el.find("rootfile")?;
     let el2 = element.borrow();
 
-    Ok(el2.get_attr("full-path")?)
-}
+    let attr = el2.get_attr("full-path")?;
 
+    Ok(PathBuf::from(attr))
+}
 
-fn build_epub_uri(path: &str, append: &str) -> String {
+fn build_epub_uri<P: AsRef<Path>>(path: P, append: &str) -> String {
     // allowing external links
     if append.starts_with("http") {
         return String::from(append);
     }
 
-    let cpath = Path::new(&path);
-    let mut cpath = cpath.to_path_buf();
+    let path = path.as_ref();
+    let mut cpath = path.to_path_buf();
 
     // current file base dir
     cpath.pop();
     for p in Path::new(append).components() {
         match p {
-            Component::ParentDir => { cpath.pop(); }
-            Component::Normal(s) => { cpath.push(s); }
+            Component::ParentDir => {
+                cpath.pop();
+            },
+            Component::Normal(s) => {
+                cpath.push(s);
+            },
             _ => {},
         };
     }
 
-    String::from("epub://") + cpath.to_str().unwrap()
+    format!("epub://{}", path.display())
```

Comment:
There's an error here, instead of path.display() here will go cpath.display(). This change will fix the broken test